src/createRouter.ts

```diff
@@ -340,13 +340,35 @@ const createRouter: RouterFactory<Context> = compose.mixin(createEvented, {
 			}
 
 			let lastDispatch: Task<void>;
+			let redirectCount = 0;
+			let redirecting = false;
+
 			const dispatch = (path: string) => {
 				if (lastDispatch) {
 					lastDispatch.cancel();
 				}
-				lastDispatch = this.dispatch(contextFactory(), path).then(({ redirect, success }) => {
+
+				// Reset redirect count if the dispatch was triggered by a non-redirect history change. This allows
+				// a route's exec / fallback / index handler to change the history, setting off a new flurry of
+				// redirects, without being encumbered by the number of redirects that led to that route being selected.
+				if (!redirecting) {
+					redirectCount = 0;
+				}
+
+				const context = contextFactory();
+				lastDispatch = this.dispatch(context, path).then(({ redirect, success }) => {
 					if (success && redirect !== undefined) {
+						redirectCount++;
+						if (redirectCount > 20) {
```

Comment:
I don't think it needs to be configurable, no. It's just there to prevent breakaway redirects. If you need more than 20 for your logic to work you're probably doing something wrong.

I considered declaring it as a constant. But it's also used only once (and admittedly in the tests). It seemed unnecessary.

Going to land this PR now but let me know if you want me to change it.